### PR TITLE
make SpringModule aware of bindings in wrapping Injector

### PR DESCRIPTION
**Pull request by twicksell**
attempted implementation for #12

### CODE REVIEW

File: src/main/java/org/springframework/guice/module/SpringModule.java [hunk 1]
```diff
@@ -24,58 +24,95 @@
...
+        this.configClasses = configClasses; 
+        this.basePackages = null;
+    }
+    
+    public SpringModule(String... basePackages) {
+        this.basePackages = basePackages;
+        this.configClasses = null;
+    }
+    
+    public SpringModule(ApplicationContext context) {
```
Comment (twicksell) [inline src/main/java/org/springframework/guice/module/SpringModule.java]: @dsyer unsure what to do with this constructor. The crux of my feature here requires the user to not pass in a refreshed ApplicationContext. Instead they can pass config classes and packages, such that I can build an AppContext myself with the required AutowireCandidateResolver. 

What do you think? I can subclass this out into another type of module, InjectorAwareSpringModule. I can deprecate this constructor, however we do use it internally in ModuleRegistryConfiguration. This functionality has been a big help to us internally, just unsure how best to package it relative to the previous implementation.

Comment (dsyer) [inline src/main/java/org/springframework/guice/module/SpringModule.java]: Hmm. I see the problem. But there's no great API to work with config classes before they are a context outside of spring boot. We don't want to depend on Boot here so we'll need to look for something cleverer. Thinking...

File: src/main/java/org/springframework/guice/module/SpringModule.java [hunk 2]
```diff
@@ -94,18 +139,55 @@ private void bindConditionally(Binder binder, String name, Class<Object> type, P
...
+	private final class PartiallyRefreshableApplicationContext extends AnnotationConfigApplicationContext {
+
+		private final AtomicBoolean partiallyRefreshed = new AtomicBoolean(false);
+
+		/*
+		 * Initializes beanFactoryPostProcessors only to ensure that all
+		 * BeanDefinition's are available
+		 */
+		void partialRefresh() {
+			invokeBeanFactoryPostProcessors(getBeanFactory());
```
Comment (twicksell) [inline src/main/java/org/springframework/guice/module/SpringModule.java]: Not happy about this approach either, but I couldn't determine any other way to get all bean definitions without doing a full refresh.

**Comment by dsyer** [comment]
Merged (and modified) in 490fc20

**Comment by twicksell** [comment]
Thanks! And no, Guice has no lifecycle management. I think the best compromise is to add a binding  either to the ApplicationContext, or preferrably some new interface we create that wraps the ApplicationContext and allows the user to close it manually. SpringGuiceContextManager.close() or something like that. That would at least give people who've implemented a Lifecycle feature on top of Guice a way to clean up the context. Governator, for example, automatically closes all bindings which implement Closeable.

**Pull request closed without merging**